### PR TITLE
Adds a new tool to calculate principal components (PCA) from a targe…

### PR DESCRIPTION
**Pull request by vruano**
…t coverage collection.

The PCA data is written into a output HDF5 file.

Example:

  java -jar hellbender-prorected.jar CalcualteCoverageComponents -I my-counts.tab -o pca.hd5

### CODE REVIEW

**Comment by vruano** [comment]
@samuelklee  please take a look.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/CalculateCoverageComponents.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            new TargetArgumentCollection(() -> this.inputFile);
+
+    @Argument(
+            doc = "PCA output file",
+            fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME,
+            optional = true
+    )
+    protected File outputFile;
+
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/CalculateCoverageComponents.java]: whitespace

File: src/main/java/org/broadinstitute/hellbender/tools/exome/CalculateCoverageComponents.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+/**
+ * Perform PCA on coverage counts.
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+@CommandLineProgramProperties(
+        summary = "Perform PCA on coverage counts",
+        oneLineSummary = "Perform PCA on coverage counts",
+        programGroup = ExomeAnalysisProgramGroup.class)
+public class CalculateCoverageComponents extends SparkToggleCommandLineProgram {
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/CalculateCoverageComponents.java]: gradle test gives warning: [serial] serializable class CalculateCoverageComponents has no definition of serialVersionUID

File: src/main/java/org/broadinstitute/hellbender/tools/exome/CalculateCoverageComponents.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+    @ArgumentCollection
+    protected TargetArgumentCollection targets =
+            new TargetArgumentCollection(() -> this.inputFile);
+
+    @Argument(
+            doc = "PCA output file",
+            fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME,
+            optional = true
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/CalculateCoverageComponents.java]: should this argument be optional?

File: src/main/java/org/broadinstitute/hellbender/tools/exome/NormalizationType.java [hunk 4]
```diff
@@ -0,0 +1 @@
+
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/NormalizationType.java]: what is this file?

File: src/main/java/org/broadinstitute/hellbender/tools/exome/TargetArgumentCollection.java [hunk 5]
```diff
@@ -58,7 +58,7 @@
      * @throws IllegalArgumentException if {@code defaultTargetFileSupplier} is {@code null}.
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/TargetArgumentCollection.java]: not sure if you wanted to change "defaultTargetFileSupplier -> defaultTargetsFileSupplier" everywhere, but you missed a few (4) cases.  also might want to change null message below accordingly.

File: src/main/java/org/broadinstitute/hellbender/utils/pca/PCA.java [hunk 6]
```diff
@@ -51,9 +77,14 @@
      *
      * @throws IllegalArgumentException if either {@code dataMatrix} or {@code svdFactory} is {@code null}.
      */
-    public static PCA createPCA(final RealMatrix dataMatrix, final Function<RealMatrix, SVD> svdFactory) {
+    public static PCA createPCA(final List<String> variables, final List<String> samples, final RealMatrix dataMatrix, final Function<RealMatrix, SVD> svdFactory) {
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/pca/PCA.java]: going forward, do we want to always require variable and sample names when doing PCA?  perhaps give the option to createPCA with no variable/sample names, and only check for names when writing to HDF5?

File: src/test/java/org/broadinstitute/hellbender/tools/exome/CalculateCoverageComponentsIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        final List<Target> targets = IntStream.range(0, testMatrix.length)
+                .mapToObj(i -> new Target("TARGET_" + i, new SimpleInterval("1", i * 100 + 1, i * 100 + 51)))
+                .collect(Collectors.toList());
+
+        try (final TableWriter<ReadCountRecord> writer = ReadCountCollectionUtils.writerWithIntervals(new FileWriter(outputFile), sampleNames)) {
+            for (int i = 0; i < testMatrix.length; i++) {
+                writer.writeRecord(new ReadCountRecord(targets.get(i), testMatrix[i]));
+            }
+        } catch (final IOException ex) {
+            throw new GATKException("", ex);
```
Comment (samuelklee) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/CalculateCoverageComponentsIntegrationTest.java]: add a message, or perhaps just fail test?

**Comment by samuelklee** [comment]
just a few minor issues.  back to you, @vruano

**Comment by vruano** [comment]
back to @samuelklee

**Comment by samuelklee** [comment]
looks good to me, just a few minor typos.  feel free to squash/merge after.  back to @vruano

**Comment by vruano** [comment]
Test fail due to some problem currently existing in master (described in #239) will merge.

**Pull request merged**